### ISSUE TITLE
ivy throw [tagName] is not a known element

### ISSUE THREAD

**Issue by elvisbegovic** [angular/angular#33685]
# 🐞 bug report

### Affected Package
I think, the issue is caused by package `@angular/core`

### Is this a regression?

Yes, it works in `ViewEngine`

### Description

In angular v9 rc1 enable ivy with certain 3th party lib throw `* is not known element`

## 🔬 Minimal Reproduction
One line repro :
`git clone https://github.com/elvisbegovic/ivy-tag-is-not-a-known-element.git && cd ivy-tag-is-not-a-known-element && npm i && ng s -o`

Compilation/serve is OK, but at runtime (open dev tools) error is:

## 🔥 Exception or Error
<pre><code>
core.js:6068 ERROR Error: 'tree-viewport' is not a known element:
1. If 'tree-viewport' is an Angular component, then verify that it is part of this module.
2. If 'tree-viewport' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@NgModule.schemas' of this component to suppress this message.
    at validateElement (core.js:23156)
    at Module.ɵɵelementStart (core.js:22963)
    at TreeComponent_Template (tree.component.js:245)
    at executeTemplate (core.js:13684)
    at renderView (core.js:13486)
    at renderComponent (core.js:14841)
    at renderChildComponents (core.js:13291)
    at renderView (core.js:13512)
    at renderComponent (core.js:14841)
    at renderChildComponents (core.js:13291)
</code></pre>

Snippet code of compiled [tree.component.js](https://gist.github.com/elvisbegovic/84a83e66e6175fc647768c53c6d77e51)

## 🌍  Your Environment

<details><summary>See Environment...</summary>

```

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-rc.1
Node: 12.13.0
OS: win32 x64
Angular: 9.0.0-rc.1
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.1
@angular-devkit/build-angular     0.900.0-rc.1
@angular-devkit/build-optimizer   0.900.0-rc.1
@angular-devkit/build-webpack     0.900.0-rc.1
@angular-devkit/core              9.0.0-rc.1
@angular-devkit/schematics        9.0.0-rc.1
@ngtools/webpack                  9.0.0-rc.1
@schematics/angular               9.0.0-rc.1
@schematics/update                0.900.0-rc.1
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
</code></pre>

```
</details>

## ✍️ Anything else relevant?
This breaks application at runtime, annoying we cannot see that durning build/serve step.

/cc @alxhub @petebacondarwin

**Comment by joshcomley** [comment]
I am having this exact same issue, but I think I can narrow it down to `<ng-content select="...">` usage. See my simple example project here (no 3rd party library, only an additional library in the workspace):

https://github.com/joshcomley/ng91tagbug

One line:

`git clone https://github.com/joshcomley/ng91tagbug.git && cd ng91tagbug && npm i && ng build mylib && ng serve`

This works fine with v9 rc0 but errors on v9 rc1.

**Comment by mattdsteele** [comment]
I'm seeing something similar when trying to use custom elements. This only regressed in `rc.1`, custom elements were working fine in `rc.0` and earlier.

To reproduce, create a new app using the CLI, add `  schemas: [CUSTOM_ELEMENTS_SCHEMA]` to the module, and add a new elements to a template, such as `<foo-bar>`. This will throw this error:

`ERROR Error: "'foo-bar' is not a known element`

Changing the `angularCompilerOption` `"enableIvy": false,` fixes the issue.

**Comment by pkozlowski-opensource** [comment]
Probably due to https://github.com/angular/angular/commit/c83f5013bfb0182dc653d7e5d8a1750e09ea50e9

**Comment by chrste90** [comment]
This also seems to break `ngx-extended-pdf-viewer`. With rc0, the library works, but with rc1 i get  `core.js:6068 ERROR Error: 'menuitem' is not a known element:`, see more details here: https://github.com/stephanrauh/ngx-extended-pdf-viewer/issues/169

**Comment by mattdsteele** [comment]
@pkozlowski-opensource thanks for the commit.

Cursory view of it makes me think https://github.com/angular/angular/commit/c83f5013bfb0182dc653d7e5d8a1750e09ea50e9 doesn't play nicely with lazy-loaded custom elements (i.e. the tags are on the page, but don't have `customElements.define()` called against them until later after they've been loaded).

This is the default behavior of web components compiled with Stencil: https://stenciljs.com/blog/how-lazy-loading-web-components-work

Was the check against `customElements.get()` occurring in the past?

**Comment by stephanrauh** [comment]
Additional information to https://github.com/angular/angular/issues/33685#issuecomment-553388338:

The validation method [`validateElement()`](https://github.com/angular/angular/blob/c83f5013bfb0182dc653d7e5d8a1750e09ea50e9/packages/core/src/render3/instructions/element.ts#L246) is elegant, but it breaks on corner cases like `menuitem`.

Currently, `menuitem` is implemented by Firefox, but not by Chrome or Safari. As far as I know, `menuitem` was part of the HTML standard before HTML5. It has been marked as deprecated, but Firefox continued to support this tag.

Be that as it may, some people - such as the pdf.js team - are using both `menu` and `menuitem` in their application. That's why ngx-extended-pdf-viewer run fine on Firefox but is broken on Chrome and Safari.

**Comment by negue** [comment]
At first I thought this issue was only after an error happened, but now this happens to me randomly on a rebuild (9.0.0-rc.2) with material components (currently only using these as 3rd party components) - the only fix is to kill and restart the `serve` process.

**Comment by elvisbegovic** [comment]
More infos: my issue with certain lib happened at first « ng serve » at runtime and only when I navigate to the lazyroute where I use lib’s directive.

And if I remove deps of the 3th party lib I can confirm what @negue is saying, I get some `* is not a kown element` when CLI rebuild.

**Comment by mhevery** [comment]
Seems related to https://github.com/angular/angular/issues/32388

**Comment by petebacondarwin** [comment]
I think the missing property is actually `directives: ...` but I agree that ngcc is not generating it for this package.

**Comment by petebacondarwin** [comment]
I have identified the root, root cause of the problem in @cexbrayat's one-line reproduction. This is probably not the general problem for other commenters on this issue but see the bottom of my analysis for ideas in those cases.

1) The root cause is actually that `angular-tree-component`'s UMD build is not valid. Their use of rollup does not specify `mobx`, `mobx-angular` and `lodash` as external packages, which leads to the UMD bundle including these packages inline.
  Note that the ESM files do not bundle up these external libraries and instead import them as expected.
2) ngcc computes dependencies between packages before processing them; it must ensure that packages are processed after their dependences. Do do this it parses just the first of the code formats that it recognises. Because it considers the package.json properties in [this order](https://github.com/angular/angular/blob/e030375d9a1e9e5b133f4c2584d5e92cdb60f222/packages/compiler-cli/ngcc/src/packages/entry_point.ts#L75-L76) - `['fesm2015', 'fesm5', 'es2015', 'esm2015', 'esm5', 'main', 'module']` it finds the `main` property (UMD) before the `module` property. So ngcc was using the broken UMD format for computing the dependencies and so missing the `mobx-angular` package.
3) Since `mobx-angular` was not found as a dependency of `angular-tree-component` the former was being processes "after" the latter, which meant that `angular-tree-component` had a missing dependency when it came to processing the ESM format (aka `module`).
4) The missing dependency during compilation can be identified as an error but is currently being swallowed to prevent [unwanted false positive failures](https://github.com/angular/angular/blob/4be8929844ae8333122584c804f18dbc9457a349/packages/compiler-cli/src/ngtsc/scope/src/local.ts#L239-L243).
5) The compilation (and so the ngcc processing) completes "successfully" but the `directives` used by the component are not generated in the rendered output, leading to the runtime errors we see here.

Possible resolutions:

1) Workaround by adding a `ngcc.config.js` that hides the `main` property in `angular-tree-component` so that we always use the `module` property for both processing and for dependency resolution.
2) Fix `angular-tree-component` - provide a PR to fix this package's distributable
3) Change the order of dependency resolution properties, so that `module` is before `main`. I think it was originally in this order because it was felt that a single flat file would be parsed faster than having to load up a deep non-flat hierarchy of files; but it is also reasonable to assume that `main` is more likely to be "broken" than `module` in 3rd party libraries out there.
4) Add better diagnostic reporting to ngtsc to help diagnose these scenarios.

**Comment by joshcomley** [comment]
So what about my example above?

**Comment by petebacondarwin** [comment]
I'll look at yours next @joshcomley  :-)

**Comment by elvisbegovic** [comment]
**I can confirm it works!!!** 

Having this temporary `ngcc.config.js` with (if I understand): 

```
module.exports = {
  packages: {
    'angular-tree-component': {
      entryPoints: {
        '.': {
          override: {
            main: undefined
          }
        }
      }
    }
  }
};
```
Or removing this line `"main": "dist/angular-tree-component.umd.js",`  in `node_modules/angular-tree-component/package.json` 

Then rerun ngcc, makes our large-app working with ivy!

By the way, doing this first `ng serve` always throw this warning:
```
WARNING in Entry point 'angular-tree-component' contains deep imports into 'C:/Users/xxx/node_modules/lodash/defaultsDeep', 'C:/Users/xxx/node_mo
dules/lodash/get', 'C:/Users/xxx/node_modules/lodash/omit', 'C:/Users/xxx/node_modules/lodash/isNumber', 'C:/Users/xxx
/node_modules/lodash/first', 'C:/Users/xxx/node_modules/lodash/last', 'C:/Users/xxx/node_modules/lodash/some', 'C:/Users/e.begovic/WebstormProjec
ts/i10/node_modules/lodash/every', 'C:/Users/xxx/node_modules/lodash/compact', 'C:/Users/xxx/node_modules/lodash/find', 'C:/Users/e.begovic/Webst
ormProjects/i10/node_modules/lodash/isString', 'C:/Users/xxx/node_modules/lodash/isFunction', 'C:/Users/xxx/node_modules/lodash/throttle', 'C:/Us
ers/xxx/node_modules/lodash/includes', 'C:/Users/xxx/node_modules/lodash/pick'. This is probably not a problem, but may cause the compilation of
entry points to be out of order.

```

Thank you @petebacondarwin

**Comment by petebacondarwin** [comment]
That warning is expected, since the `angular-tree-component` imports deeply from `lodash` which are not "proper" APF entry-points. But it doesn't matter in this case because `lodash` is not an Angular library.

**Comment by petebacondarwin** [comment]
With the change to the diagnostics reporting we get:

```
Compiling angular-tree-component : module as esm5
console_logger.ts:38
Error: Failed to compile entry-point angular-tree-component due to compilation errors:
node_modules/@angular/common/common.d.ts(115,22): error TS-996002: Appears in the NgModule.imports of TreeModule, but could not be resolved to an NgModule class
node_modules/mobx-angular/dist/mobx-angular.d.ts(5,22): error TS-996002: Appears in the NgModule.imports of TreeModule, but could not be resolved to an NgModule class
```

which at least will tell us where to start looking and not leave us with an obscure runtime error.

**Comment by petebacondarwin** [comment]
@joshcomley - as far as I can tell your scenario is working as expected. The `<myselector>` element is not known to Angular - it is neither a well-known HTML element nor a component/directive defined with an element selector.

Note also that this has nothing to do with `MyLibComponent` being in a separate package (or even `NgModule`). You can move it to the main package (and declare it in the `AppModule` and you get the same result.  It is possible that this error was not shown in JIT mode until recently when we landed https://github.com/angular/angular/commit/c83f5013bfb0182dc653d7e5d8a1750e09ea50e9.

**Comment by petebacondarwin** [comment]
I have created a new issue for the `NO_ERRORS_SCHEMA` scenario which is different to this issue: https://github.com/angular/angular/issues/33976

**Comment by petebacondarwin** [comment]
Closing this as ngcc is actually working as expected and we have landed the PR to improve the error messaging.

**Comment by stephanrauh** [comment]
@petebacondarwin Just doublechecking: did you address the `menuitem` issue, too? 

IMHO that's a tricky one. It works on Firefox but fails on most other browsers.

**Comment by petebacondarwin** [comment]
@stephanrauh - your case is similar to @joshcomley 's case IMO, so let's move the discussion there: https://github.com/angular/angular/issues/33976
